### ISSUE TITLE
[Enhancement] Round corners set to zero issue

### ISSUE THREAD

**Issue by coatsy35** [Courseplay/Courseplay_FS22#3384]
**READ FIRST**
https://github.com/Courseplay/Courseplay_FS22/wiki

**Steps to reproduce**
Field 30 on Vintage Oaks

**Actual behavior**
If you select sharpen corners on a field with 3 headlands and set the number of round corners to 0 you get two rounded corners then one sharp corner. If you set the rounded corners to 1, you then get one rounded corner and two sharp corners. See images below

**Expected behavior**
0 for round corners should force all corners to be sharp, ideally going off in a virtual line to create a virtual "90 degree" corner so the corner can be pulled out straight.

**Screenshots**
![fsScreen_2024_09_01_22_55_53](https://github.com/user-attachments/assets/184279be-fcdb-4b91-be65-701afb72cb6d)
![fsScreen_2024_09_01_22_55_12](https://github.com/user-attachments/assets/5a377adc-fcc3-431f-a35c-e1c44786cbe6)

![fsScreen_2024_09_01_22_56_16](https://github.com/user-attachments/assets/73add28c-a325-45ca-9e65-5032837b095b)
![fsScreen_2024_09_01_22_56_10](https://github.com/user-attachments/assets/95fbdf7c-ebf2-48e0-885e-e65ff38ab50e)

**Courseplay version**
Version used: 7.4.2.0

**Log**
https://drive.google.com/file/d/1Vw4reNx1F9_zwLrtajUH6741PCzfnddZ/view?usp=drive_link

**Map**
Vintage Oaks

**Vehicles/ implements involved**
Which vehicle or implement?
Mod yes/no ?

**Relevant script mods**
Mods like Autodrive or VCA and so on.

**Comment by pvaiko** [comment]
The generator only sharpens/rounds corners which have a radius less than the vehicle's turning radius. Corners, such as the one in your screen shot, already round with a radius above the vehicle's radius aren't changed. 

If the expectation is sharpen these two, then we need to introduce another parameter as the generator will need to know what is the radius limit, under which sharpening should happen.

**Comment by coatsy35** [comment]
> The generator only sharpens/rounds corners which have a radius less than the vehicle's turning radius. Corners, such as the one in your screen shot, already round with a radius above the vehicle's radius aren't changed.
> 
> If the expectation is sharpen these too, then we need to introduce another parameter as the generator will need to know what is the radius limit, under which sharpening should happen.

I did think that might be the case of needing a parameter for it - would 0 on rounded corners be the basis for the new logic? That would be a great enhancement in time now you have got it all working so well.

PS. How do you know your vehicle turning radius? is there a place to check what it is for a mod or Giants mahcine?

**Comment by Tensuko** [comment]
I would love a more automatic logical system more, like the radius is always lower then vehicle turn radius or something.
The point is how to explain to ppl what to do.... will figure something out.

About Radius, there is a command for that, don't know wich right now.
For controlling the radius with a setting in vehicle menu, that must wait untill we have something for proper control it.
(Read value and never evern can set lower then the read value to make sure it always makes the turns and stuff).

**Comment by rohne83** [comment]
Don't understand the issue here. In the screenshots everything works as it should work. There's no possibility to make the outermost corner sharp as the field shape just does not allow it. Vehicle would have to go off the field here for what coatsy wants to have. If he wants that corner calculated sharp also for the outermost lane, he has to amend the field shape first by himself anyway (ploughing or with some script mods he can even do this with cultivators or sowing machines). Don't think it would be a good behaviour of CP course generator to go off the field shape due to some settings where people don't have such effect in mind.

**Comment by pvaiko** [comment]
>I rather see issues with multitool courses in this kind of corners (imagine that corner with 3 tools and just one headland, instead of the 1 tool with 3 headlands).

With the new multitool implementation, if you select 1 round corner, only the vehicle working on the outermost headland will make a round corner.

**Comment by rohne83** [comment]
> > I rather see issues with multitool courses in this kind of corners (imagine that corner with 3 tools and just one headland, instead of the 1 tool with 3 headlands).
> 
> With the new multitool implementation, if you select 1 round corner, only the vehicle working on the outermost headland will make a round corner.

Really? That's great. Just have to wait until the general multitool bug is fixed then...

**Comment by coatsy35** [comment]
> Don't understand the issue here. In the screenshots everything works as it should work. There's no possibility to make the outermost corner sharp as the field shape just does not allow it. Vehicle would have to go off the field here for what coatsy wants to have. If he wants that corner calculated sharp also for the outermost lane, he has to amend the field shape first by himself anyway (ploughing or with some script mods he can even do this with cultivators or sowing machines). Don't think it would be a good behaviour of CP course generator to go off the field shape due to some settings where people don't have such effect in mind. As long as the radius is big enough to not leave empty spaces, the automatic decision for sharp vs round is the most appropriate way to go and so the courses above are totally correct in my eyes. I rather see issues with multitool courses in this kind of corners (imagine that corner with 3 tools and just one headland, instead of the 1 tool with 3 headlands).

@rohne83 @pvaiko 

You would never want any form of short curve if you were ploughing (And I know we can't replicate everything as in IRL) but I have ploughed 1000's of acres and you always make the corner square unless it's a nice long smooth curve) I create a virtual line out of the corner in my head and head to it, the same with drilling actually or any other tool that doesn't like to follow curves unless they are very long and smooth such as tedding, spraying, fertiliser spreading. 

In my image above the option of 0 needs removing really then, as currently it actually gives your two round corners if you pay attention to the image when set at 0. If you set it to 1 it works correctly and gives you 1 round corner.

**Comment by rohne83** [comment]
Yes, but as already said: if you want such corners to be completely sharp out until the last row, you should make that edge sharp manually before by yourself. CP will detect this and generate the course as you want. But it would never be a good behaviour of the course generator to generate a course going off the current field boundaries just because you set the number of round corners to 0.

**Comment by coatsy35** [comment]
> Yes, but as already said: if you want such corners to be completely sharp out until the last row, you should make that edge sharp manually before by yourself. CP will detect this and generate the course as you want. But it would never be a good behaviour of the course generator to generate a course going off the current field boundaries just because you set the number of round corners to 0.

You know, real GPS steering units have a feature called Square corners which do as I said, project a line out of the field. With an automated approach you would need to detect when the nose of the machine touches the field border and then reverse up. The part under your tractor not covered, is then covered by your next pass backing the implement up the same, to touch the field border.

Nice video here explaining it for anyone who cares to know how they work

https://youtu.be/psFpdSOZCQ0?si=nHdngzGsRetOOfso

**Comment by pvaiko** [comment]
As I said, the generator can do this, there is just no setting for it on the UI, like this corner (this a screenshot from out test tool):
![image](https://github.com/user-attachments/assets/6ec79c0a-4483-4349-bf14-70c2143149dc)

**Comment by coatsy35** [comment]
> As I said, the generator can do this, there is just no setting for it on the UI, like this corner (this a screenshot from out test tool): ![image](https://private-user-images.githubusercontent.com/2379521/366199447-6ec79c0a-4483-4349-bf14-70c2143149dc.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjYwMDAzNjcsIm5iZiI6MTcyNjAwMDA2NywicGF0aCI6Ii8yMzc5NTIxLzM2NjE5OTQ0Ny02ZWM3OWMwYS00NDgzLTQzNDktYmYxNC03MGMyMTQzMTQ5ZGMucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDkxMCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA5MTBUMjAyNzQ3WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9N2E1NWU3Nzk0M2E5Zjc1MDBjMmMwM2QzYjUxMDMzMTdiY2Y0N2E4MDM0ODlkNDEyYjIxMTFjNGIyNjJmNmEzNyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.C50wnoCU7fKGlDUygBmasz3diQeSEWRhVfG1d7Do7lo)

Now that is cool! Wow. Really hope you decide to put a setting in for it at some point.

**Comment by Tensuko** [comment]
Every setting makes the generator more complex.
Some ppl already reported to me that they think its too many options but they just need to get used to it.
So I am 100% against adding another setting.

The question is, if we just turn that value all the way down to make sure:
Sharp corner is sharp corner, no matter what.

Since if you don't want sharp corners, you just turn it off and have that "very tiny rounded version".
For rounder corners, there is the round corner option.

So basically all situations that make sence can be done already, just that sharp at some point isn't always sharp.
That should be adressed if we can but without an unnecessary setting.

**Comment by coatsy35** [comment]
Surely the too many options issue is addressed with the advanced vs non advanced options? Or what about having an external options file for those that like more control that can be turned on if required and most people would never mess with?

It would make sense to me as you say for that to not have another option but make the 0 round corners force sharp corners for all? Then if it doesn’t suit, you can just increase the number not altering anything else for people and keeping it simple still.

**Comment by rohne83** [comment]
> So basically all situations that make sence can be done already, just that sharp at some point isn't always sharp. That should be adressed if we can but without an unnecessary setting.

But what will happen when you have several types of corners on same field, which is the case quite often. So you have corners as the one that coatsy showed in his screenshots, but also really sharp corners. Most players would want the latter to be worked sharp indeed but for the one which coatsy showed stick to the current behavior with the tool following the contour (and rather amend the field size manually before course generation if you still want it that corner to be worked sharp as well). Especially when you also have obstacles near the field edge it would be counterproductive if the tool went outside the defined field definition.

**Comment by Tensuko** [comment]
In FS with all its different shaped fields and tools, the question "what if..." Has endless answers.
There is a point where we have to stop and say thats enough possibilitys.
Obstacles near field thats why we have round corners counted from 1st headland.

**Comment by coatsy35** [comment]
For interests sake, JD have just released this video of their new software for field boundaries and row. Which shows what I was describing. 

https://youtu.be/FTC8CPzvDAg?si=QSidxpdls-sETB4T&t=17
